sos/utilities.py

```diff
@@ -496,8 +562,45 @@ def running(self):
         return self.process.poll() is None
 
 
-class AsyncReader(threading.Thread):
-    """Used to limit command output to a given size without deadlocking
+class HeadReader(threading.Thread):
+    """Used to 'head' the command output (f_src) to a given size
+    without deadlocking sos. Takes a sizelimit value in MB.
+    """
+
+    COPY_BUFSIZE = 1024*1024
+
+    def __init__(self, f_src, f_dst, sizelimit, binary):
+        super().__init__()
+        self.f_src = f_src
+        self.f_dst = f_dst
+        self.remaining = sizelimit * 1048576  # convert to bytes
+        self.binary = binary
+        self.running = True
+        self.start()
+
+    def run(self):
+        """Reads from the f_src (Popen stdout pipe) until we reach sizelimit.
+        once done, close f_src to signal the program that we are done.
+        """
+        while self.remaining > 0:
+            buf = self.f_src.read(min(self.remaining, self.COPY_BUFSIZE))
+            if not buf:
+                break
+            self.f_dst.write(buf)
+            self.remaining -= len(buf)
+        self.f_src.close()
+        self.running = False
+
+    def get_contents(self):
+        return '' if not self.binary else b''
```

Comment:
note rather to myself: we return empty string here, despite the method is called when we return status 124, 126 or 127..? (I need to review behaviour in these use cases).